### PR TITLE
Assemble oc and kubectl helper runner

### PR DESCRIPTION
**Pull request by prietyc123**
**What type of PR is this?**

/kind code-refactoring

**What does does this PR do / why we need it**:

```helper.cli.go``` will have common functions for ```kubectl``` and ```oc``` and executes the function according to the cluster type, hence it will ease the work for pr related to issue : https://github.com/openshift/odo/issues/2752

**Which issue(s) this PR fixes**:

Fixes NA

**How to test changes / Special notes to the reviewer**:

All the test which includes ```kubectl```  and ```oc``` command should run.

### CODE REVIEW

**Comment by prietyc123** [comment]
prow failure

/retest

File: tests/helper/helper_cluster.go [hunk 1]
```diff
@@ -13,36 +13,36 @@ import (
...
+	// path to cluster binary
 	path string
 }
 
-// NewOcRunner initializes new OcRunner
-func NewOcRunner(ocPath string) OcRunner {
-	return OcRunner{
-		path: ocPath,
+// NewClusterRunner initializes new ClusterRunner
+func NewClusterRunner(clusterPath string) ClusterRunner {
```
Comment (amitkrout) [inline tests/helper/helper_cluster.go:22]: Literally ClusterRunner looks like you are provisioning or running a cluster through the script, isn't it ? I would suggest a better phrase for what exactly you want to do, for example ```func NewClirRunner(clusterPath string) CliRunner {```. I mean replace ```ClusterRunner``` with ```CliRunner```

@prietyc123 Does it make sense ?

File: tests/helper/helper_cluster.go [hunk 2]
```diff
@@ -13,36 +13,36 @@ import (
...
+// Run cluster with given arguments
+func (cluster *ClusterRunner) Run(cmd string) *gexec.Session {
 	session := CmdRunner(cmd)
 	Eventually(session).Should(gexec.Exit(0))
 	return session
 }
 
 // SwitchProject switch to the project
-func (oc *OcRunner) SwitchProject(projectName string) {
+func (cluster *ClusterRunner) SwitchProject(projectName string) {
```
Comment (amitkrout) [inline tests/helper/helper_cluster.go:36]: replace with something like this ```func (oc *CliRunner) SwitchProject(projectName string) {``` because this is an OpenShift specific function and it won't be supported for ```kubectl``` command. Find out all those type of function and update accordingly. @prietyc123

**Comment by girishramnani** [comment]
/hold
waiting for @cdrage's comments to be resolved

File: tests/helper/helper_cluster.go (hunk 2 above)
Comment (prietyc123) [reply tests/helper/helper_cluster.go:36]: Yes, we can replace ```cluster``` with ```cli``` but if we try to replace specifically go-lint will throw error as ```receiver name oc should be consistent with previous receiver name cli for CliRunner go-lint```

File: tests/helper/helper_cluster.go [hunk 3]
```diff
@@ -13,36 +13,36 @@ import (
 	"github.com/onsi/gomega/gexec"
 )
 
-type OcRunner struct {
-	// path to oc binary
+type CliRunner struct {
```
Comment (kadel) [inline tests/helper/helper_cluster.go:16]: Why renaming this to generic `CliRunner`? 
There are functions in this that are specific just for `oc`.

File: tests/helper/helper_cluster.go (hunk 1 above)
Comment (prietyc123) [reply tests/helper/helper_cluster.go]: Done

File: tests/helper/helper_cluster.go (hunk 3 above)
Comment (prietyc123) [reply tests/helper/helper_cluster.go:16]: We have some test where we have to switch runner according to the cluster type i.e Kubernetes or openshift. In that case handling the kubectl and oc runner will become tough. So, thinking from that prospective and reusability of code I was thinking to generalise it for kubectl and oc. 

Yup, like you mentioned function like SwitchProject... won't work for sure, however we can handle those through if...else statement or may be through switch statment depending upon which client is being used. May be in another issue i can analysze those incompatibility and fix those in a separate pr.

**Comment by amitkrout** [comment]
> /hold
> waiting for @cdrage's comments to be resolved

@girishramnani Are you sure ? Is it @cdrage or @amitkrout ? ;)

**Comment by prietyc123** [comment]
prow failure

/retest

File: tests/helper/helper_cluster.go (hunk 3 above)
Comment (girishramnani) [reply tests/helper/helper_cluster.go]: I would suggest we don’t rename to CliRunner

**Comment by girishramnani** [comment]
> > /hold
> > waiting for @cdrage's comments to be resolved
> 
> @girishramnani Are you sure ? Is it @cdrage or @amitkrout ? ;)
Yeah @amitkrout’s. Got mixed.

File: tests/helper/helper_cluster.go (hunk 3 above)
Comment (prietyc123) [reply tests/helper/helper_cluster.go]: ok, let me revisit the change again, also let me find a better way to implement it. Till then i am holding the pr

Comment (amitkrout) [reply tests/helper/helper_cluster.go]: > ok, let me revisit the change again, also let me find a better way to implement it. Till then i am holding the pr

+1

**Comment by prietyc123** [comment]
Thanks @mik-dass  for helping with this PR.

File: tests/helper/helper_cliRunner.go [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package helper
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/onsi/gomega/gexec"
```
Comment (mik-dass) [inline tests/helper/helper_cliRunner.go:6]: remove these unused imports

File: vendor/github.com/onsi/gomega/go.mod [hunk 5]
```diff
@@ -1,17 +0,0 @@
-module github.com/onsi/gomega
-
-require (
```
Comment (mik-dass) [inline vendor/github.com/onsi/gomega/go.mod:3]: why was this removed?

Comment (prietyc123) [reply vendor/github.com/onsi/gomega/go.mod]: Thanks for pointing it out. It got removed accidently.

File: tests/helper/helper_cliRunner.go (hunk 4 above)
Comment (prietyc123) [reply tests/helper/helper_cliRunner.go]: > remove the above unused imports

Done

**Comment by prietyc123** [comment]
prow failure

/retest

File: tests/helper/helper_cliRunner.go (hunk 4 above)
Comment (prietyc123) [reply tests/helper/helper_cliRunner.go]: done

File: tests/helper/helper_cluster.go (hunk 3 above)
Comment (prietyc123) [reply tests/helper/helper_cluster.go]: > I would suggest we don’t rename to CliRunner

@girishramnani @kadel I have done some modification with common interface for all the cli. Can you please look into this once.

**Comment by prietyc123** [comment]
prow failure

/retest

**Comment by prietyc123** [comment]
We are hitting timeout error in the [test script](https://github.com/openshift/odo/blob/master/tests/integration/servicecatalog/cmd_service_test.go#L311)

```
Running odo with args [odo create nodejs frontend --context /tmp/134975824 --project mmdynenmau]
[odo] Validation
[odo]  •  Validating component  ...
 ✓  Validating component [6ms]
[odo] 
[odo] Please use `odo push` command to create the component with source deployed
Running odo with args [odo push --context /tmp/134975824]
[odo] Validation
[odo]  •  Checking component  ...
 ✓  Checking component [118ms]
[odo] 
[odo] Configuration changes
[odo]  ✓  Initializing component
[odo]  •  Creating component  ...
 ✓  Creating component [324ms]
[odo] 
[odo] Pushing to component frontend of type local
[odo]  •  Checking files for pushing  ...
 ✓  Checking files for pushing [1ms]
[odo]  •  Waiting for component to start  ...
[odo]  ✗  Waiting for component to start [4m]
[odo]  ✗  waited 4m0s but couldn't find running pod matching selector: 'deploymentconfig=frontend-app'
Setting current dir to: /home/travis/gopath/src/github.com/openshift/odo/tests/integration/servicecatalog
Deleting project: mmdynenmau
Running odo with args [odo project delete mmdynenmau -f]
[odo] This project contains the following applications, which will be deleted
[odo] Application app
[odo] This application has following components that will be deleted
[odo] component named frontend
[odo]  •  Deleting project mmdynenmau  ...
 ✓  Deleting project mmdynenmau [6s]
[odo]  ✓  Deleted project : mmdynenmau
Deleting dir: /tmp/856178805
Deleting dir: /tmp/134975824
Deleting dir: /tmp/742594671
• Failure [249.631 seconds]
odo service command tests
/home/travis/gopath/src/github.com/openshift/odo/tests/integration/servicecatalog/cmd_service_test.go:15
  When link backend between component and service
  /home/travis/gopath/src/github.com/openshift/odo/tests/integration/servicecatalog/cmd_service_test.go:301
    should link backend to service successfully [It]
    /home/travis/gopath/src/github.com/openshift/odo/tests/integration/servicecatalog/cmd_service_test.go:308
    No future change is possible.  Bailing out early after 241.099s.
    Running odo with args [odo push --context /tmp/134975824]
    Expected
        <int>: 1
    to match exit code:
        <int>: 0
    /home/travis/gopath/src/github.com/openshift/odo/tests/helper/helper_run.go:34
```

**Comment by prietyc123** [comment]
Issue is already tracked https://github.com/openshift/odo/issues/2088#issuecomment-618263407 

/retest

**Comment by prietyc123** [comment]
Prow failure 
```
level=info msg="Waiting up to 30m0s for the cluster at https://api.ci-op-m6htyzp6-18180.origin-ci-int-aws.dev.rhcloud.com:6443 to initialize..."
level=fatal msg="failed to initialize the cluster: Multiple errors are preventing progress:\n* Could not update servicemonitor \"openshift-apiserver-operator/openshift-apiserver-operator\" (430 of 435): the server does not recognize this resource, check extension API servers\n* Could not update servicemonitor \"openshift-authentication-operator/authentication-operator\" (389 of 435): the server does not recognize this resource, check extension API servers\n* Could not update servicemonitor \"openshift-cluster-version/cluster-version-operator\" (8 of 435): the server does not recognize this resource, check extension API servers\n* Could not update servicemonitor \"openshift-console-operator/console-operator\" (406 of 435): the server does not recognize this resource, check extension API servers\n* Could not update servicemonitor \"openshift-controller-manager-operator/openshift-controller-manager-operator\" (434 of 435): the server does not recognize this resource, check extension API servers\n* Could not update servicemonitor \"openshift-image-registry/image-registry\" (395 of 435): the server does not recognize this resource, check extension API servers\n* Could not update servicemonitor \"openshift-kube-apiserver-operator/kube-apiserver-operator\" (409 of 435): the server does not recognize this resource, check extension API servers\n* Could not update servicemonitor \"openshift-kube-controller-manager-operator/kube-controller-manager-operator\" (414 of 435): the server does not recognize this resource, check extension API servers\n* Could not update servicemonitor \"openshift-kube-scheduler-operator/kube-scheduler-operator\" (420 of 435): the server does not recognize this resource, check extension API servers\n* Could not update servicemonitor \"openshift-machine-api/cluster-autoscaler-operator\" (153 of 435): the server does not recognize this resource, check extension API servers\n* Could not update servicemonitor \"openshift-machine-api/machine-api-operator\" (422 of 435): the server does not recognize this resource, check extension API servers\n* Could not update servicemonitor \"openshift-operator-lifecycle-manager/olm-operator\" (424 of 435): the server does not recognize this resource, check extension API servers\n* Could not update servicemonitor \"openshift-service-catalog-apiserver-operator/openshift-service-catalog-apiserver-operator\" (398 of 435): the server does not recognize this resource, check extension API servers\n* Could not update servicemonitor \"openshift-service-catalog-controller-manager-operator/openshift-service-catalog-controller-manager-operator\" (401 of 435): the server does not recognize this resource, check extension API servers"
```

/retest

**Comment by amitkrout** [comment]
> lgtm

/lgtm
